### DISCUSSION TITLE
Optuna for diffusion parameters optimization

### DISCUSSION THREAD

**Discussion by abrahami** [optuna/optuna#5548] [Q&A]
Hi,
I have a problem of finding the optimal parametrs for a **network** structure. The network consists of nodes (users) and edges (interactions between users). Edges have a fixed probability value of an interaction between the users (beta).
I know the seed users who are infected (at time t0) and the ones who are infected by the end of the process (t1).
I would like to find the optimal beta of this diffusion process. Can (and should) I use optuna for that?
If so, how exactly?
Thank you!

**Comment by toshihikoyanase** [comment]
Hi @abrahami ,

Thank you for your question.
Could you provide us with more details about your problem settings?

> The network consists of nodes (users) and edges (interactions between users). 

Can we assume that the number of users remains fixed in each experiment?
If so, do we need to determine the value of each element in the adjacency matrix (user by user)?

> I would like to find the optimal beta of this diffusion process.

How do you evaluate the different beta values quantitatively?
In other words, could you share the metric that you are using in your experiment?

**Reply by abrahami** [reply]
thanks @toshihikoyanase!
Please let me give some extra information as you suggested.
The network is fixed -- the number of nodes and edges do not change and are known in advance. Of course some users do not have an edge between them. Those that have an edge, have some probability to interact.

We need to determine the value of each element in the adjacency matrix. However, in the simplest format, only a single value exist over all existing edges, and we want to find the best one to fit the data. Of course, later we want to try a more complex scenario with multiple parameters to optimize (e.g., more complex structures of the network in which different type of edges exist).

The data we have in hand are the users that were infected at t=0 and the ones infected by the end of the process. We assume that a few iterations of interactions between users occur in between, but we do not have this information of the users who got infected in between.

We evaluate the performance of an optimization algorithm, by comparing the number of infected users -- true VS algorithm prediction. For simplicity, we compare the performance by the root square of the error -- sqrt(#predicted_infected - #true_indected)^2)

If Optuna is a good fit for this problem type, we will test it. I have not seen an example of Optuna handling this type of data source, but if it works it can be a great scenario example for the usage of the package.

**Reply by toshihikoyanase** [reply]
Thank you for the detailed information, @abrahami .

I believe we can use Optuna to tackle your problem. Here's an example flow for applying Optuna.

> only a single value exist over all existing edges, and we want to find the best one to fit the data.

Since you are initially looking for a single beta value, you can explore the value by using `trial.suggest_float` as follows:

```python
def objective(trial):
    ...
    # Please update the range to your problem setting.
    beta = trial.suggest_float("beta", 0, 1)
```

Next, simulate the diffusion process to obtain the number of infected users and compute the performance metric:

```python
def objective(trial):
    ...
    # Suggest beta
    beta = trial.suggest_float("beta", 0, 1)
    ...
    # Simulate infection process
    n_predict = infecton_simulation(beta, graph, infected_t0, infected_t1)
    # Compute and return the metric.
    return math.sqrt((n_predict - n_true) ** 2)
```

Finally, please create a study and start optimization:

```python
import optuna

study = optuna.create_study(direction="minimize")
study.optimize(objective, n_trials=100)
print(f'Best trial: {study.best_trial.value}, Best beta: {study.best_trial.params["beta"]}')
```

> I have not seen an example of Optuna handling this type of data source, but if it works it can be a great scenario example for the usage of the package.

Sounds interesting. If the above flow suits your problem, please share your insights.

**Reply by abrahami** [reply]
Thank you @toshihikoyanase !
Your input is really helpful and the code snippets are great. I was able to run a test with a toy example and find the accurate single "beta" parameter.

Next I wish to apply it over a multi parameters optimization problem. This will be based on a _multilayer_ network. Each layer has its own beta parameter (inner edge probability). In addition, edges _between_ layers exist, so infection can move over the layers. In this setup, each pair of layers has its own "gamma" parameter infection probability.

Is there an example of how to apply the package for such a multi parameters problem?
Thank you!

**Reply by toshihikoyanase** [reply]
Thank you for your follow-up question @abrahami. Optuna definitely supports multi-parameter problems.

Suppose we have `num_layers` layers, we can add `suggest_float` calls corresponding to `beta` parameters and `gamma` parameters as follows:

```python
def objective(trial):
    betas = [trial.suggest_float(f"beta_{i}", 0, 1) for i in range(num_layers)]
    gammas = []
    for i in range(num_layers):
        for j in range(num_layers):
            if i == j:
                continue
            gammas.append(trial.suggest_float(f"gamma_{i}_{j}", 0, 1))
```

Note that as we add more parameters, the optimization difficulty will increase. Therefore, it might be a good idea to start small.

**Reply by abrahami** [reply]
Thank you @toshihikoyanase !
Indeed, we shall start small to see how it progresses. I was able to implement it with two parameters, and now we'll move forward with trying it using multiple parameters in the actual setup.
Please let us update you here if we find Optuna a good solution for the use case, and if so, we can add it to the repo examples or write a short blog about it

Thanks again!

**Reply by toshihikoyanase** [reply]
Thank you for the reply @abrahami!

I'm glad to hear that you started to implement the code for your experiments.
Looking forward to hearing your updates!

**Reply by abrahami** [reply]
Hi @toshihikoyanase again!
We have made much progress using Optuna to optimize multiple hyperparameters for a network structure problem.
We are able to find the best parameters while optimizing some objective function.
However, we still have an open question that we are struggling with, and hopefully, you can shed some light on.

When we look at the objective function **over iterations** (i.e., steps), we do not see that the objective value converges. In other words, we see that the objective is very noisy and jumps from value to value while it does not converge.
Is this behavior reasonable? Is this expected from such a (multiple parameters) optimization problem?
We are currently using the default optimization algorithm (_TPESampler_)

Thanks a lot!
